homeassistant/components/recorder/queries.py

```diff
@@ -621,3 +621,22 @@ def find_statistics_runs_to_purge(
 def find_latest_statistics_runs_run_id() -> StatementLambdaElement:
     """Find the latest statistics_runs run_id."""
     return lambda_stmt(lambda: select(func.max(StatisticsRuns.run_id)))
+
+
+def find_legacy_event_state_and_attributes_and_data_ids_to_purge(
+    purge_before: datetime,
+) -> StatementLambdaElement:
+    """Find the latest row in the legacy format to purge."""
+    return lambda_stmt(
+        lambda: select(
+            Events.event_id, Events.data_id, States.state_id, States.attributes_id
+        )
+        .join(States, Events.event_id == States.event_id)
```

Comment:
Its using a `.join` here instead of an `.outerjoin` so as soon as there are none left that have an event_id, we switch over the new method.